### DISCUSSION TITLE
Native Numbering for Cross-Refs in DOCX

### DISCUSSION THREAD

**Discussion by iandol** [quarto-dev/quarto-cli#2464] [Feature Requests]
At present, Quarto uses plain text for cross-referenced output to DOCX. DOCX does support native cross-ref numbering, and the ideal situation would be for Quarto cross references to generate native numbers. This has been requested for other cross-ref filters before: `pandoc-crossref` doesn't generate native figure numbers, I made an issue for this but the developer does not want to add output-specific code: https://github.com/lierdakil/pandoc-crossref/issues/299 - another crossref plugin has a similar request and a bit more progress: https://github.com/tomduck/pandoc-fignos/issues/34 

I recently found **this has been solved** by an rmarkdown project: https://github.com/davidgohel/officedown — they apparently create native numbers and also therefore support list of figures/tables. I think this would improve the Quarto goal of supporting as many features cross-format as possible... They offer some other nice features like supporting table and list styles (I'm not an rmarkdown user so haven't tested it). Surely this R code can help inspire a Lua filter solution?

DOCX, unfortunately, remains a dominant format for many scientific journal submissions, and also the common format working with many colleagues who aren't code nerds or don't understand what markdown is...

**Comment by iandol** [comment]
Note that Pandoc has partially solved this natively (`native_numbering`), though it appears there are still some problems:

https://github.com/jgm/pandoc/issues/7392

Note I also posted this idea previously but it disappeared when I searched and I resubmitted, see https://github.com/quarto-dev/quarto-cli/discussions/2335 for the first post...

**Reply by acxz** [reply]
The feature in pandoc works just as expected.

**Comment by iusgit** [comment]
Hi,
it seems I need this, too.....

In https://github.com/jgm/pandoc/issues/7392 the pandoc extension `native_numbering`mentioned, that one can call with `-t docx+native_numbering` in pandoc standalone.

How can I select that in quarto? I tried:
```
format:
  docx:
    native_numbering: true
```
but didn't succeed.

Any ideas?

**Reply by acxz** [reply]
Pandoc extensions are treated as format variants in Quarto, use like so:
```quarto
format:
  docx:
    variant: +native_numbering
```

**Comment by iandol** [comment]
AFAIK Pandoc's native numbering is probably incompatible with Quarto's cross-referencing, or at least I have no idea how these two could play together. I am hoping the Pandoc code can inform an update to Quarto's cross-referencing lua filters.

In the meantime to get LOF or LOT you can use the Word tools to search for the caption styles for figures and tables to make the LOF. e.g.

<img width="273" alt="Screenshot 2022-10-04 at 19 33 13" src="https://user-images.githubusercontent.com/153905/193810030-164ce20a-8ab0-4a9d-8bb0-5048ee242b96.png">

This will fail if you are using Quarto's multipart figures or tables (where each part of a figure gets added, then the whole figure legend too) and you'll need some manual tweaking...

**Reply by iusgit** [reply]
As discussed here (https://github.com/iandol/scrivomatic/issues/48#issuecomment-1266770605) one can add openxml syntax to the document to get LOF and LOT. Lua filter might be necessary if one wants to define the exact position in the document.

**Reply by acxz** [reply]
@iusgit can you be specific and outline how to do so, preferably with an example?

**Reply by acxz** [reply]
For additional information, here is the issue in pandoc for creating lof and lot for docx: https://github.com/jgm/pandoc/issues/8245

**Reply by iandol** [reply]
See this sample document, I just took the XML that pandoc generates for TOC, and modified it for LOF, both raw chunks are here:

````markdown
---
title: hello
author: Jane Doe
---

```{=openxml}
<w:sdt>
<w:sdtPr><w:docPartObj><w:docPartGallery w:val="Table of Contents" /><w:docPartUnique /></w:docPartObj></w:sdtPr>
<w:sdtContent>
<w:p><w:pPr><w:pStyle w:val="TOCHeading" /></w:pPr><w:r><w:t xml:space="preserve">Table of Contents</w:t></w:r></w:p>
<w:p><w:r><w:fldChar w:fldCharType="begin" w:dirty="true" />
<w:instrText xml:space="preserve">TOC \o "1-3" \h \z \u</w:instrText>
<w:fldChar w:fldCharType="separate" /><w:fldChar w:fldCharType="end" /></w:r></w:p>
</w:sdtContent>
</w:sdt>
```

```{=openxml}
<w:sdt>
<w:sdtPr><w:docPartObj><w:docPartGallery w:val="Table of Figures" /><w:docPartUnique /></w:docPartObj></w:sdtPr>
<w:sdtContent>
<w:p><w:pPr><w:pStyle w:val="TOCHeading" /></w:pPr><w:r><w:t xml:space="preserve">Table of Figures</w:t></w:r></w:p>
<w:p><w:r><w:fldChar w:fldCharType="begin" w:dirty="true" />
<w:instrText xml:space="preserve">TOC \h \z \t "Image Caption" \c</w:instrText>
<w:fldChar w:fldCharType="separate" /><w:fldChar w:fldCharType="end" /></w:r></w:p>
</w:sdtContent>
</w:sdt>
```

# A

Blah

![Image A](placeholder.png)

# B

Blah

![Image B](placeholder.png)
````

Gives:
<img width="313" alt="image" src="https://github.com/user-attachments/assets/b6ae9541-c0de-42b2-b9ff-a7008aa2fbc5">

You can perhaps pare down the XML to a simpler set of tags, but Word is fussy and it needs trial and error...

**Reply by acxz** [reply]
@iandol thanks for the quick reply unfortunately, the sample code you posted does not work for me.

I think there might have been an error during your copy/paste (see third to last line). The second list of figures block should be:

````pandoc
```{=openxml}
<w:sdt>
<w:sdtPr><w:docPartObj><w:docPartGallery w:val="Table of Figures" /><w:docPartUnique /></w:docPartObj></w:sdtPr>
<w:sdtContent>
<w:p><w:pPr><w:pStyle w:val="TOCHeading" /></w:pPr><w:r><w:t xml:space="preserve">Table of Figures</w:t></w:r></w:p>
<w:p><w:r><w:fldChar w:fldCharType="begin" w:dirty="true" />
<w:instrText xml:space="preserve">TOC \h \z \t "Image Caption" \c</w:instrText>
<w:fldChar w:fldCharType="separate" /><w:fldChar w:fldCharType="end" /></w:r></w:p>
</w:sdtContent>
</w:sdt>
```
````

After this fix, it worked for me. Thanks so much! btw: if you could edit your comment with the fix that would be great!

**Reply by iandol** [reply]
Fixed, thanks and glad it works. By the way you can remove the some of the xml, basically leave the 5-7 lines and the TOC should still work.
